### ISSUE TITLE
llvm --with-clang --with-lldb fails to build

### ISSUE THREAD

**Issue by aw1621107** [Homebrew/homebrew#42041]
[gist](https://gist.github.com/aw1621107/bd44249b2d70311f6e00)

Seems to fail with codesign: identity not found or something like that?

Definitely sure I have the lldb_codesign certificate installed, and fairly certain I've done it correctly (or at least I've done it in a way that worked in the past on 10.10)

**Comment by DomT4** [comment]
```
[ 92%] Building CXX object tools/lldb/source/Plugins/SymbolFile/DWARF/CMakeFiles/lldbPluginSymbolFileDWARF.dir/DWARFDeclContext.cpp.o
cd /tmp/llvm20150722-32717-1hh1jjt/tools/lldb/source/Plugins/SymbolFile/DWARF && /usr/local/Library/ENV/4.3/clang++   -DGTEST_HAS_RTTI=0 -DHAVE_ROUND -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fPIC -fvisibility-inlines-hidden -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wcovered-switch-default -Wnon-virtual-dtor -std=c++11 -Wno-deprecated-declarations -Wno-unknown-pragmas -Wno-deprecated-register  -fno-exceptions -fno-rtti -fPIC -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -I/tmp/llvm20150722-32717-1hh1jjt/tools/lldb/source/Plugins/SymbolFile/DWARF -I/tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/tools/lldb/source/Plugins/SymbolFile/DWARF -I/tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/tools/lldb/include -I/tmp/llvm20150722-32717-1hh1jjt/tools/lldb/include -I/tmp/llvm20150722-32717-1hh1jjt/include -I/tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/include -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/python2.7 -I/tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/tools/lldb/../clang/include -I/tmp/llvm20150722-32717-1hh1jjt/tools/lldb/../clang/include -I/tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/tools/lldb/source/.    -UNDEBUG -fno-exceptions -fno-rtti -o CMakeFiles/lldbPluginSymbolFileDWARF.dir/DWARFDeclContext.cpp.o -c /tmp/llvm20150722-32717-qu1p4p/llvm-3.6.1.src/tools/lldb/source/Plugins/SymbolFile/DWARF/DWARFDeclContext.cpp
cd /tmp/llvm20150722-32717-1hh1jjt/bin && CODESIGN_ALLOCATE=/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate codesign --force --sign lldb_codesign debugserver
lldb_codesign: no identity found
```

The identity does exist, otherwise you wouldn't have gotten this far into compile. Did you follow all the instructions from https://llvm.org/svn/llvm-project/lldb/trunk/docs/code-signing.txt?

**Comment by aw1621107** [comment]
I'm fairly sure I did. Tried it multiple times too, so hopefully I didn't mess up every single time... Shows up under `certtool y` too, so not really sure why it can't be found

**Comment by DomT4** [comment]
Try codesigning something with it:

```
touch /tmp/test.txt
codesign -s lldb_codesign /tmp/test.txt
codesign -vvvv /tmp/test.txt
```

**Comment by aw1621107** [comment]
```
awang-rp% cd ~/Desktop
awang-rp% touch test.txt
awang-rp% codesign -s lldb_codesign test.txt
awang-rp% codesign -vvvv test.txt
test.txt: valid on disk
test.txt: satisfies its Designated Requirement
awang-rp%
```

**Comment by DomT4** [comment]
Strange, that definitely seems fine.

Try killing the identity and following https://llvm.org/svn/llvm-project/lldb/trunk/docs/code-signing.txt through again. It's forever possible Apple changed something to do with certificate handling on 10.11, but the process has been reported to work fine on 10.9-10.10 with the LLVM instructions.

`Always Trust` and `Certificate Type = Code Signing` are particularly important.

**Comment by aw1621107** [comment]
Already did that (before posting the issue, didn't seem to do anything. Will do it again though.

Also, should I be seeing both a public and private key for the certificate? I only see a private one, but the gdb certificate I have has both public and private...

**Comment by DomT4** [comment]
I'll try and stick my head in my 10.11 VM either tonight or tomorrow evening and see if Apple has done something with the process on 10.11.

**Comment by aw1621107** [comment]
Alright, it's building now. I'll update when it succeeds/fails.

Few notes about the key creation process:

I couldn't drag the certificate from keychain to keychain; the icon just vanished as soon as it moved over one of the keychain icons. I manually copy/pasted instead. Don't know if that works exactly the same, but it seems to move everything visible over.

After following the steps, I have the certificate + public key in the login keychain, and the private key is in both login and system keychains?

**Comment by aw1621107** [comment]
Nope, failed with same error.

**Comment by AlexanderOMara** [comment]
I don't think this is a 10.11 issue. I'm getting the exact same error on 10.10.4.

Here's [a Gist](https://gist.github.com/AlexanderOMara/de7e20687b67a40c527b) from my compilation using the following command:

    HOMEBREW_MAKE_JOBS=1 brew install -v llvm --universal --with-clang --with-lld --with-lldb --with-python --with-rtti 2>&1

I'm using XCode 6.4. Perhaps it's an issue with the newer XCode?

**Comment by AlexanderOMara** [comment]
Interestingly enough, I tried running the command that is failing on a test Mach-O binary (just a simple hello world), and it succeeded.

    cd ~/Downloads/ && CODESIGN_ALLOCATE=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate codesign --force --sign lldb_codesign test

**Comment by aw1621107** [comment]
Waaaaaaaait a second...
02.make:8206
```
cd /tmp/llvm20150727-562-jlnv37/tools/lldb/tools/debugserver/source/MacOSX && mig -I/tmp/llvm20150727-562-jlnv37/tools/lldb/tools/debugserver/source/MacOSX /tmp/llvm20150727-562-10vt89b/llvm-3.6.1.src/tools/lldb/tools/debugserver/source/MacOSX/dbgnub-mig.defs
/tmp/llvm20150727-562-10vt89b/llvm-3.6.1.src/tools/lldb/tools/debugserver/source/MacOSX/dbgnub-mig.defs:13:10: fatal error: 'DNBConfig.h' file not found
#include "DNBConfig.h"
         ^
1 error generated.
```
Maybe this killed the build instead? So possibly #39621 again?

**Comment by aw1621107** [comment]
Also seems to pop up [here](https://github.com/Keno/Cxx.jl/issues/4). Possible fix?

**Comment by xu-cheng** [comment]
I can reproduce the same error on 10.10.4

**Comment by DomT4** [comment]
The `#include "DNBConfig.h"` error is the same one we were seeing before, but we've never been able to consistently identify what causes it to not be created in certain circumstances but created in others.

I'm wondering if it's related to whether people have Xcode, the CLT, or both installed, but that's just a complete theory.

**Comment by AlexanderOMara** [comment]
@DomT4 Would the `#include "DNBConfig.h"` error be the cause of the `lldb_codesign: no identity found` error?

I have installed both XCode and the CLT.

**Comment by aw1621107** [comment]
I believe I had just XCode the first few times, and installed the (rest of the?) CLT tools for the second try

**Comment by DomT4** [comment]
> Would the #include "DNBConfig.h" error be the cause of the lldb_codesign: no identity found error?

It's possible one erroring out would somehow prevent successful execution of the codesigning step. Given compile calls the `DNBConfig.h` error fatal I'm surprised it continues on as much as it does. [This is apparently the perl script intended to generate the necessary header](https://opensource.apple.com/source/lldb/lldb-179.1/tools/debugserver/scripts/dbgnub-config.pl).

**Comment by aw1621107** [comment]
Hmm... Weird thing is that `DNBConfig.h` *is* being generated; it's just not being found.

**Comment by zchee** [comment]
Build faild too.
but, faild point is `[ 97%] Built target lldbTarget`.

More infomation, see [https://gist.github.com/zchee/3cfd446e9aafb7f0443f](https://gist.github.com/zchee/3cfd446e9aafb7f0443f)

**NOTE:** I want to `lldb python bindings` linking the Brewed Python.
So, I added
```
-DPYTHON_INCLUDE_DIR='/usr/local/Cellar/python/HEAD/Frameworks/Python.framework/Versions/2.7/include/python2.7'
-DPYTHON_LIBRARY='/usr/local/Cellar/python/HEAD/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib'
```

**Comment by DomT4** [comment]
>  I want to lldb python bindings linking the Brewed Python.

`brew install llvm --with-python` would have worked for that.

Same overall error though: ` fatal error: 'DNBConfig.h' file not found `.

**Comment by zchee** [comment]
@DomT4 Sorry, In my case, When the PC that has been upgraded from yosemite in el capitan, the flag did not work properly.
I'll try to test in vmware.

**Comment by DomT4** [comment]
> the flag did not work properly.

The `:python` one?

**Comment by kylecook80** [comment]
I am having the same issue as well. Here is my output: https://gist.github.com/kylecook80/01a4ff3ede8ca3357502

**Comment by aw1621107** [comment]
Hmmm...

Seems [a recent commit (11 Aug 2015)](http://reviews.llvm.org/D11971) wiped out DNBConfig.h and anything to do with it. `--HEAD` builds [still seem to fail](https://gist.github.com/aw1621107/e53101ab6a20a9aa0b44) though, but at least it's not the missing header breaking things this time (or it isn't showing up in the logs).

The last command before the build fails is
```
cd /tmp/llvm20150831-63297-khzanh/bin && CODESIGN_ALLOCATE=/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate codesign --force --sign lldb_codesign debugserver
```
Running the second part of the line on some random text file and a spare compiled executable I had laying around didn't result in any errors, so maybe something in the build environment is blocking access to the keychain?

**Comment by aw1621107** [comment]
Yep... The second part of the above line works fine in my regular shell, but fails to find lldb_codesign in the debug shell created once the build fails.

Also, [this](http://stackoverflow.com/a/16038076/3580294) Stack Overflow answer seems to shed some light on the issue. Running `certtool y` in my regular shell yields a list of certificates/info, while running the same in the debug shell results in `***Error retrieving default keychain`. `certtool y v` yields `***Error retrieving default keychain : SecKeychainCopyDefault: A default keychain could not be found.`

Trying `security default-keychain -s <login.keychain>` didn't change anything either.

**Comment by zchee** [comment]
@DomT4 Is this problem is not yet solved?
I with to solve the problem, and I can solve even python problems.
Since the build is successful without a homebrew, I will send a pull request if you are a necessary.

**Comment by aw1621107** [comment]
Alright, got another piece of the puzzle.

Seems the login keychain can only be accessed through a login shell, and the debug shell doesn't appear to be a login shell. If the llvm build is running in a non-login environment, that would explain the missing identity.

```
awang-rp% certtool y v
***Error retreiving default keychain : SecKeychainCopyDefault: A default keychain could not be found.
awang-rp% login
login: awang
Password:
Last login: Wed Sep  2 14:15:25 on ttys004
awang-rp% certtool y v
    <list of certificates>
awang-rp% logout
awang-rp% certtool y v
***Error retreiving default keychain : SecKeychainCopyDefault: A default keychain could not be found.
awang-rp%
```

**Comment by zchee** [comment]
@DomT4 ping. and llvm updated `3.7.0`.

**Comment by RandomDSdevel** [comment]
@DomT4:  Does LLVM v3.7.0 contain a fix for this?

**Comment by DomT4** [comment]
I haven't had much time to look at this lately. I'm wrapping up my current uni module, so time is tight, and on top of that we've had some tech issues taking precedence.

@aw1621107; You've done the most work on this, Thank you for that. Which shell have you invoked at the time of build, out of curiosity?

It looks like LLVM have actually updated the instructions for the 3.7.0 release so I'm curious if the new instructions are adapted for El Cap's changes: https://llvm.org/svn/llvm-project/lldb/trunk/docs/code-signing.txt

**Comment by aw1621107** [comment]
I was in a zsh login shell when I ran the brew command.

Unfortunately, seems the new release didn't really change
much. --HEAD still fails to build, albeit for a different reason:

```
gmake[6]: *** No rule to make target 'lib/libLLVMSupport.a', needed by 'bin/clang-tblgen'. Stop.
```

I have a feeling I saw this earlier, but don't remember
what I did with it... I'll take a closer look when I get home

**Comment by aw1621107** [comment]
Hmmm...
```
...
[100%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/SystemUtils.cpp.o
[100%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/ToolOutputFile.cpp.o
[100%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/TargetParser.cpp.o
[125%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Triple.cpp.o
[125%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Timer.cpp.o
...
```

**Comment by RandomDSdevel** [comment]
Wait, _what?_  That just looks…_wrong!_  Ugh, what's causing _that?_

**Comment by aw1621107** [comment]
And now it's failing "normally". Yay reproducibly

**Comment by aw1621107** [comment]
Seems the issue might not be login/non-login shells. Still tied to something about the session, though.

Output from a normal terminal window:
```
awang-rp% [[ -o login ]] && echo 'login' || echo 'not login'
login
awang-rp% security default-keychain
    "/Users/awang/Library/Keychains/login.keychain"
awang-rp% zsh
awang-rp% [[ -o login ]] && echo 'login' || echo 'not login'
not login
awang-rp% security default-keychain
    "/Users/awang/Library/Keychains/login.keychain"
```

Output from the debug shell:
```
awang-rp% [[ -o login ]] && echo 'login' || echo 'not login'
not login
awang-rp% security default-keychain
security: SecKeychainCopyDefault: A default keychain could not be found.
```

**Comment by aw1621107** [comment]
All right, think I got it! (or at least something that seems like it works)

From the debug shell:
```
awang-rp% cd /tmp
awang-rp% touch test.txt
awang-rp% codesign -s lldb_codesign /tmp/test.txt
lldb_codesign: no identity found
awang-rp% export HOME=/Users/awang
awang-rp% codesign -s lldb_codesign /tmp/test.txt
awang-rp% export HOME=/private/tmp/llvm20150909-40636-1j81c4r/.brew_home
awang-rp% codesign -s lldb_codesign /tmp/test.txt
lldb_codesign: no identity found
```

So seems like `codesign` is depending on `$HOME` to find the keychain.
Homebrew changes that to a temp directory, and `codesign` doesn't know what to do any more.

@DomT4 Any way to set `$HOME` for that command and/or surrounding commands? Or any other alternatives?

**Comment by aw1621107** [comment]
Unfortunately, seems a keychain tweak won't work:
```
awang-rp% security list-keychains
    "/Library/Keychains/System.keychain"
awang-rp% security list-keychains -s /Users/awang/Library/Keychains/login.keychain
awang-rp% security list-keychains
    "/Library/Keychains/System.keychain"
```

**Comment by aw1621107** [comment]
Oh wait
```
awang-rp% security list-keychains
    "/Library/Keychains/System.keychain"
awang-rp% sudo security list-keychains -d common -s /Users/awang/Library/Keychains/login.keychain
awang-rp% security list-keychains
    "/Users/awang/Library/Keychains/login.keychain"
```

Not sure if homebrew has control during the right part of the process though...

**Comment by aw1621107** [comment]
.....
```
[125%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Watchdog.cpp.o
[125%] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Watchdog.cpp.o
[125%] Linking CXX static library ../libLLVMSupport.a
[125%] Linking CXX static library ../libLLVMSupport.a
/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: ../libLLVMSupport.a(oUTF32) /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive member: ../libLLVMSupport.a(oUTF32) size too large (archive member extends past the end of the file)
size too large (archive member extends past the end of the file)
/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar: internal ranlib command failed
lib/Support/CMakeFiles/LLVMSupport.dir/build.make:2460: recipe for target 'lib/libLLVMSupport.a' failed
gmake[6]: *** [lib/libLLVMSupport.a] Error 1
```

Different build error every time...

**Comment by aw1621107** [comment]
Alright, both stable and --HEAD built!

I ran something like this:
```
awang-rp% security list-keychains -d common
<something I don't know>
awang-rp% security list-keychains -d common -s /Users/awang/Library/Keychains/login.keychain
awang-rp% brew install llvm --with-clang --with-lldb --debug
awang-rp% security list-keychains -d common -s <original keychain>
```

So that at least takes care of this issue. I'm testing with other flags, but I'd guess different flags would be a different issue...

**Comment by aw1621107** [comment]
Er, didn't want to click that...

**Comment by aw1621107** [comment]
Got a build `--with-clang --with-lld --with-lldb --with-python` working.
Had to start another session inside of the normal one though; kept
getting random build errors until then, or it just happened to work
that time around. Not sure...

**Comment by xu-cheng** [comment]
@aw1621107 Thanks for exploring and testing. You are right by moving all of `lldb_codesign` keys to system from login solves the problem for me.

**Comment by RandomDSdevel** [comment]
Doesn't part of LLVM's guide on how to code-sign LLDB instruct you to do something like that?

**Comment by aw1621107** [comment]
It does. I believe gdb's instructions tell/let you leave its certificate in the system keychain

**Comment by RandomDSdevel** [comment]
Huh.  The LLDB guide reads in a manner that suggests that you only have to leave a _copy_ of your code-signing certificate in the System keychain and can also leave the original in your login keychain.

**Comment by aw1621107** [comment]
Hmmm...
```
awang-rp% security list-keychains
    "/Library/Keychains/System.keychain"
    "/private/tmp/llvm20150911-26429-1mxuomu/.brew_home/Library/Keychains/login.keychain"
```
No idea why it decides to do this this time and not earlier times, but seems this fix may not be consistent

**Comment by DomT4** [comment]
Oh. Heh. We override the `HOME` environmental variable for security & sandbox reasons. It's possible we need to give LLVM some leeway on that.

**Comment by aw1621107** [comment]
I wonder... Any way to set `$HOME` just for that command? Maybe patching a (c)makefile to run `HOME=/Users/____ codesign ...` instead?

**Comment by aw1621107** [comment]
Also looks like the error with >100% goes away if you drop into shell, `make clean` and `make`. I'll need to look at that too...

**Comment by dongcarl** [comment]
Still having trouble here: https://gist.github.com/dongcarl/44315070ec232dd56cc8
ran `brew install llvm --with-clang --with-lld --with-lldb --with-rtti`

**Comment by aw1621107** [comment]
@dongcarl This issue should be resolved if/when #46721 is merged

**Comment by RandomDSdevel** [comment]
@aw1621107:  Awesome, and thanks for putting the work into this!
